scipy/special/special/tools.h

```diff
@@ -0,0 +1,77 @@
+/* Building blocks for implementing special functions */
+
+#pragma once
+
+#include "config.h"
+#include "error.h"
+
+namespace special {
+namespace detail {
+
+    // Return NaN, handling both real and complex types.
+    template <typename T>
+    SPECFUN_HOST_DEVICE inline std::enable_if_t<std::is_floating_point_v<T>, T> maybe_complex_NaN() {
+        return std::numeric_limits<T>::quiet_NaN();
+    }
+
+    template <typename T>
+    SPECFUN_HOST_DEVICE inline std::enable_if_t<!std::is_floating_point_v<T>, T> maybe_complex_NaN() {
+        using V = typename T::value_type;
+        return {std::numeric_limits<V>::quiet_NaN(), std::numeric_limits<V>::quiet_NaN()};
+    }
+
+    // Series evaluators.
+    template <typename Generator, typename T>
+    SPECFUN_HOST_DEVICE T series_eval(Generator &g, T init_val, double tol, std::uint64_t max_terms,
+                                      const char *func_name) {
+        /* Sum an infinite series to a given precision.
+         *
+         * g : a generator of terms for the series.
+         *
+         * init_val : A starting value that terms are added to. This argument determines the
+         *     type of the result.
+         *
+         * tol : relative tolerance for stopping criterion.
+         *
+         * max_terms : The maximum number of terms to add before giving up and declaring
+         *     non-convergence.
+         *
+         * func_name : The name of the function within SciPy where this call to series_eval
+         *     will ultimately be used. This is needed to pass to set_error in case
+         *     of non-convergence.
+         */
+        T result = init_val;
+        T term, previous;
+        for (std::uint64_t i = 0; i < max_terms; ++i) {
+            term = g();
+            result += term;
+            if (std::abs(term) < std::abs(result) * tol) {
+                return result;
+            }
+        }
+        // Exceeded max terms without converging. Return NaN.
+        set_error(func_name, SF_ERROR_NO_RESULT, NULL);
+        return maybe_complex_NaN<T>();
+    }
+
+    template <typename Generator, typename T>
+    SPECFUN_HOST_DEVICE T series_eval_fixed_length(Generator &g, T init_val, std::uint64_t num_terms) {
+        /* Sum a fixed number of terms from a series.
+         *
+         * g : a generator of terms for the series.
+         *
+         * init_val : A starting value that terms are added to. This argument determines the
+         *     type of the result.
+         *
+         * max_terms : The number of terms from the series to sum.
+         *
+         */
+        T result = init_val;
+        for (std::uint64_t i = 0; i < num_terms; ++i) {
+            result += g();
+        }
+        return result;
+    }
+
```

Comment:
These tools are awesome, I really like the generality of them. Looking forward to building them up.